Fix part of #14187: covering remaining frontend code files and migrating question-update.service from angular to angular 2+

Hatilar420 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #14187 .
2. This PR does the following: migrating `question-update.service.ts` to angular 2+ and increasing its test coverage to 100 %.

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

- video in develop branch


https://user-images.githubusercontent.com/29446639/158819704-9eb50d14-92ea-401b-9e96-1a7e649dcc64.mp4



<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly. Please also include videos/screenshots of the developer tools browser console, so that we can be sure that there are no console errors.
If the PR changes the UI in any of the files listed in rtl_css.py (i.e, those that have a separate .rtl.css file for styling), make sure to add screenshots with the site language set to Arabic as well.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

- video after migration


https://user-images.githubusercontent.com/29446639/158820266-33c9ad73-9739-47ab-bf80-8e2a679852cf.mp4

- 100% test coverage

![test_coverage_for_question_update](https://user-images.githubusercontent.com/29446639/158821268-b4f64fb4-c461-4db9-9b26-24c5ef4cdde9.png)


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Hatilar420 committed:
not updated

Hatilar420 committed:
migration to angular and 100% test coverage

Hatilar420 committed:
develop merge

Hatilar420 committed:
Merge branch 'develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
lint fixes

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

kevintab95 reviewed (changes requested):
General question about this PR, how are we achieving full coverage for `question-update.service`? I don't see any new tests being added in the relevant spec file..

Hatilar420 commented:
Hi @kevintab95

Thanks for reviewing the code.

I found no need to add new tests to the spec file , since there were only few lines in the private methods of the code that weren't getting covered. which i covered by adding few changes in the `sampleStateDict` and `expectedOutputStateDict` and by changing some code in the `updateFunction` in `should update the state data of the question` test.

Should i add a new test to the file ?

Hatilar420 commented:
@kevintab95 PTAL

kevintab95 commented:
Yep, please add relevant tests that test specific methods / functionality rather than relying on the mock data to cover stuff (which seems like a "side effect".)

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

nithusha21 commented on core/templates/domain/question/question-update.service.spec.ts:
```diff
@@ -16,86 +16,31 @@
-  'components/question-directives/question-editor/' +
-  'question-editor.component.ts');
+import { TestBed } from '@angular/core/testing';
+import {QuestionUpdateService} from 'domain/question/question-update.service';
```
Please keep the formatting consistent. Spaces after { and before }... Here and elsewhere.

Hatilar420 replied:
done

nithusha21 commented on core/templates/domain/question/question-update.service.spec.ts:
```diff
@@ -16,86 +16,31 @@
-  var QuestionObjectFactory = null;
-  var QuestionUndoRedoService = null;
-  var StateObjectFactory = null;
+  var questionUpdateService = null;
```
Use let instead of var. And provide a datatype instead of null. For example:
let questionUpdateService: QuestionUpdateService

Hatilar420 replied:
done

nithusha21 commented on core/templates/services/UpgradedServices.ts:
```diff
@@ -331,6 +331,8 @@ import { ProfilePageBackendApiService } from
 import { PythonProgramTokenizer } from 'classifiers/python-program.tokenizer';
 import { QuestionBackendApiService } from
   'domain/question/question-backend-api.service';
+import { QuestionUpdateService } from
```
I don't believe we need to add to this file anymore. @srijanreddy98 to confirm

nithusha21 commented:
Also please note the failing e2e tests. I suspect there might be change detection related issues post migration where you'll have to explicitly do a $rootscope.$apply on non-Angular components for the changes to reflect.

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
Added a new test

Hatilar420 committed:
added rootScope apply

Hatilar420 commented:
@kevintab95 @nithusha21 PTAL
1. @kevintab95   I added a new test for the coverage 
2. @nithusha21  I added `$rootScope.apply()` where the service is being used
Please take a look and let me know if PR needs any other change.
Thanks

nithusha21 commented:
Please try to fix the failing tests. There seems to be more issues.

kevintab95 commented on core/templates/components/question-directives/question-editor/question-editor.component.ts:
```diff
@@ -112,6 +112,7 @@ angular.module('oppia').component('questionEditor', {
         }
         QuestionUpdateService.setQuestionStateData(
           ctrl.question, updateFunction);
+        $rootScope.$apply();
```
Prefer `$rootScope.$applyAsync` here and elsewhere.

kevintab95 replied:
Also, I don't understand why it is added here. Could you please explain?

Hatilar420 replied:
I added them as some tests were failing. Removed them.

kevintab95 commented on core/templates/components/question-directives/question-editor/question-editor.component.ts:
```diff
@@ -179,6 +180,7 @@ angular.module('oppia').component('questionEditor', {
           angular.copy(displayedValue));
         QuestionUpdateService.setQuestionInapplicableSkillMisconceptionIds(
           ctrl.question, displayedValue);
+        $rootScope.$apply();
```
Ditto. Why is it added in this line?

kevintab95 commented on core/templates/domain/question/question-update.service.spec.ts:
```diff
@@ -16,88 +16,38 @@
+    stateObjectFactory = TestBed.inject(StateObjectFactory);
 
-    sampleStateDict = {
+    sampleStateDict1 = {
```
Avoid naming variables such that there is a trailing number that uniquely identifies it. Instead, try to give more meaningful names.

kevintab95 replied:
Ditto below.

Hatilar420 replied:
done

Hatilar420 replied:
done

kevintab95 reviewed (changes requested):
Thanks @Hatilar420! Left some comments. PTAL

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
removed apply

Hatilar420 commented:
@nithusha21 PTAL  i tried adding `$rootScope.apply()` and `$rootScope.ApplyAsync()` both of which resulted in frontend tests failing. can you please help ?

Hatilar420 commented:
@kevintab95 PTAL

kevintab95 commented on core/templates/domain/question/question-update.service.spec.ts:
```diff
@@ -16,88 +16,38 @@
+    stateObjectFactory = TestBed.inject(StateObjectFactory);
 
-    sampleStateDict = {
+    contentStateSampleStateDict = {
```
It looks like "state" is repeated twice here. Is that intentional?

kevintab95 replied:
Would `sampleQuestionStateDict` make sense?

kevintab95 commented on core/templates/domain/question/question-update.service.spec.ts:
```diff
@@ -159,7 +109,7 @@ describe('Question update service', function() {
     };
 
-    expectedOutputStateDict = {
+    expectedContentOutputStateDict = {
```
Unsure what content is referring to here..

kevintab95 commented on core/templates/domain/question/question-update.service.ts:
```diff
@@ -16,165 +16,191 @@
-      var changeObj = new Change(changeDict, apply, reverse);
-      QuestionUndoRedoService.applyChange(changeObj, question);
-    };
+  private applyChange(
```
Here and below, prefer prefixing `_` for private functions as it was before (see [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide#javascript).)

kevintab95 commented:
@Hatilar420 -- I'm a bit confused about the video uploaded in the "proof of changes" section. The changes are related to the question update service but in the test video, it seems like the exploration editor is being tested. Should we be testing the question editor instead?

kevintab95 reviewed (changes requested):
Thanks @Hatilar420! Left some comments PTAL

nithusha21 commented:
Can you run one of the failing tests locally and watch as it plays through the test? That should indicate what action fails and then you can verify if explicit change detection needs to be triggered there. The scenario you should be looking for is an AngularJS component/service using an Angular service resulting in the need for an explicit change detection cycle.

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
Fixes

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update

Hatilar420 committed:
Merge branch 'oppia:develop' into frontend_test_and_migration_for_question_update
